src/providers/twitch/eventsub/Controller.cpp

```diff
@@ -154,133 +222,157 @@ SubscriptionHandle Controller::subscribe(const SubscriptionRequest &request)
+                        });
+                        return;
+
+                    case Error::Conflict:
+                        // This session ID is already subscribed to this request, some logic of ours is wrong
+                        qCWarning(LOG) << "Conflict" << errorString << request;
+                        break;
+
+                    case Error::Ratelimited:
+                        qCDebug(LOG) << "Ratelimited" << errorString << request;
+                        break;
+
+                    case Error::Forwarded:
+                    default:
+                        qCWarning(LOG) << "Unhandled error, retrying "
+                                          "subscription"
+                                       << errorString << request;
+                        boost::asio::post(
+                            this->ioContext, [this, request]() mutable {
                                 this->retrySubscription(
                                     request, boost::posix_time::seconds(2), 5);
                             });
-                            break;
-                    }
-                });
-            return;
+                        return;
+                }
+
+                this->markRequestFailed(request);
+            });
+
+        return;
+    }
+
+    if (openButNotReadyConnections == 0)
+    {
+        // No connection was available to handle this subscription request, create a new connection
+        this->createConnection();
+        this->retrySubscription(request, boost::posix_time::millisec(500), 10);
+    }
+    else
+    {
+        if (openButNotReadyConnections > 1)
+        {
+            qCWarning(LOG) << "We have" << openButNotReadyConnections
+                           << "open but no ready connections, hmmm";
+        }
+
+        // At least one connection is open, but it has not gotten the welcome message yet
+        this->retrySubscription(request, boost::posix_time::millisec(250), 10);
+    }
+    // });
+}
+
+std::optional<std::shared_ptr<lib::Session>> Controller::getViableConnection(
+    uint32_t &openButNotReadyConnections)
+{
+    for (const auto &weakConnection : this->connections)
+    {
+        auto connection = weakConnection.lock();
+        if (!connection)
+        {
+            // TODO: remove it here?
+            continue;
         }
 
-        if (openButNotReadyConnections == 0)
+        auto *listener = dynamic_cast<Connection *>(connection->getListener());
+
+        if (listener == nullptr)
         {
-            // No connection was available to handle this subscription request, create a new connection
-            this->createConnection();
-            this->retrySubscription(request, boost::posix_time::millisec(500),
-                                    10);
+            // something really goofy is going on
+            qCWarning(LOG) << "listener was not the correct type";
+            continue;
```

Comment:
Here we continue - I think we should assert here as well.

Reply:
Fixed in dc3330b0b1267ef7e1ca416f87906fa26d9f3503